### PR TITLE
add E2E tests for AzureManagedCluster webhooks

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR adds E2E tests for the full set of webhook update rules for AKS. We don't fully test every invalid permutation across all properties (or properties of properties, i.e., structs), but rather include a working "negative" E2E test case for every `AzureManagedCluster`, `AzureManagedControlPlane`, and `AzureManagedMachinePool` webhook.

This process uncovered an unnecessary webhook for the `AzureManagedControlPlane.Name` property — the `.Name` field is the unique ID for these k8s Object.Meta resources, and updates to `.Name` violate rules that prevent the webhooks from ever being reached (in practice I observed `409` conflict errors). That webhook (and its related UT) are removed in this PR.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by jackfrancis** [comment]
@nojnhuh how does this initial E2E scenario look as a surface area for lots more webhook tests?

File: test/e2e/aks_webhooks.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	err = mgmtClient.Get(ctx, types.NamespacedName{
+		Namespace: input.Cluster.Namespace,
+		Name:      input.Cluster.Name,
+	}, amc)
+
+	By("Validating that AzureManagedCluster webhooks are working")
+	By("Validating that we can't manually add an annotation to AzureManagedCluster with a prefix of 'infrastructure.cluster.x-k8s.io/custom-header-'")
+	amc.ObjectMeta.Annotations[azure.CustomHeaderPrefix+"foo"] = "bar"
+	err = mgmtClient.Update(ctx, amc)
+	Expect(err).To(HaveOccurred())
```
Comment (nojnhuh) [inline test/e2e/aks_webhooks.go:65]: I think it would be worth doing some sort of minimal "this is an error from the webhook" validation here to be safe, like [`apierrors.IsForbidden`](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsForbidden) (assuming that's correct). I've definitely made mistakes when creating tests that generate HTTP 409 conflicts or similarly unexpected errors so I'd personally appreciate a little bit of guardrail here.

**Review by nojnhuh — COMMENTED** [review]
Left one small comment. Overall looks good!

File: test/e2e/azure_logcollector.go [hunk 2]
```diff
@@ -177,61 +176,6 @@ func getHostname(m *clusterv1.Machine, isWindows bool) string {
 	return hostname
 }
 
-func getAzureCluster(ctx context.Context, managementClusterClient client.Client, namespace, name string) (*infrav1.AzureCluster, error) {
```
Comment (jackfrancis) [inline test/e2e/azure_logcollector.go:180]: I was able to re-use one of these existing funcs, but needed another one that wasn't there, so took the opportunity to add that (`getAzureManagedCluster`) in the more general-purpose `aks.go` source file rather than here. And because of the re-use of `getAzureManagedControlPlane` it seems like we can move all of these (reusable) funcs to `aks.go` from this log collection-specific source file.

File: test/e2e/aks_webhooks.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if to.String(amcp.Spec.NetworkPolicy) == "azure" {
+		amcp.Spec.NetworkPolicy = to.StringPtr("calico")
+	} else {
+		amcp.Spec.NetworkPolicy = to.StringPtr("azure")
+	}
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
```
Comment (nojnhuh) [inline test/e2e/aks_webhooks.go:144]: I wonder if a lot of the boilerplate could be removed with an extra layer of `BeforeEach`/`AfterEach`, if Ginkgo allows that here. I imagine that might be able to turn most of this into something like:

```go
	var amcp *infrav1exp.AzureManagedControlPlane

	BeforeEach(func() {
		amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
		Expect(err).NotTo(HaveOccurred())
	})

	AfterEach(func() {
		err = mgmtClient.Update(ctx, amcp)
		Expect(err).To(HaveOccurred())
		Expect(apierrors.IsInvalid(err)).To(BeTrue())
	})

	It("fails when updating Spec.NetworkPlugin", func() {
		if to.String(amcp.Spec.NetworkPlugin) == "azure" {
			amcp.Spec.NetworkPlugin = to.StringPtr("kubenet")
		} else {
			amcp.Spec.NetworkPlugin = to.StringPtr("azure")
		}
	})
```

Certainly not necessary to address now, just an idea.

File: test/e2e/aks_webhooks.go [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+		}
+		ammp.Spec.AvailabilityZones = append(ammp.Spec.AvailabilityZones, "99")
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if len(ammp.Spec.AvailabilityZones) > 0 {
+			ammp.Spec.AvailabilityZones = ammp.Spec.AvailabilityZones[:len(ammp.Spec.AvailabilityZones)-1]
```
Comment (nojnhuh) [inline test/e2e/aks_webhooks.go:291]: If all we want to accomplish here is changing the slice, this would be a bit more concise:
```suggestion
			ammp.Spec.AvailabilityZones = ammp.Spec.AvailabilityZones[1:]
```

File: test/e2e/aks_webhooks.go [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't add or remove from AzureManagedMachinePool.Spec.AvailabilityZones")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if ammp.Spec.AvailabilityZones == nil {
+			ammp.Spec.AvailabilityZones = []string{}
+		}
+		ammp.Spec.AvailabilityZones = append(ammp.Spec.AvailabilityZones, "99")
```
Comment (nojnhuh) [inline test/e2e/aks_webhooks.go:283]: FWIW I think this nil check can be omitted since appending to a nil slice directly works the same way: https://go.dev/play/p/m5rP1TWFiDU

File: test/e2e/aks_webhooks.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.EnableUltraSSD")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if to.Bool(ammp.Spec.EnableUltraSSD) {
+			ammp.Spec.EnableUltraSSD = to.BoolPtr(false)
+		} else {
+			ammp.Spec.EnableUltraSSD = to.BoolPtr(true)
+		}
```
Comment (nojnhuh) [inline test/e2e/aks_webhooks.go:336]: Maybe a bit too clever, but:
```suggestion
		ammp.Spec.EnableUltraSSD = to.BoolPtr(!to.Bool(ammp.Spec.EnableUltraSSD))
```

File: test/e2e/aks_webhooks.go [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+		} else {
+			ammp.Spec.KubeletConfig = &infrav1exp.KubeletConfig{
+				FailSwapOn: to.BoolPtr(true),
+			}
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By(fmt.Sprintf("Changing AzureManagedMachinePool %s mode to User", ammp.Name))
```
Comment (nojnhuh) [inline test/e2e/aks_webhooks.go:379]: I can see where if the cluster template ever includes more than one System node pool, then this test would have a side effect of changing all but one of them to User node pools. I think the likelihood of the template changing that way and that causing any actual issues is low enough that I doubt it's worth addressing that now though.

File: test/e2e/azure_test.go [hunk 8]
```diff
@@ -589,6 +589,16 @@ var _ = Describe("Workload cluster creation", func() {
...
+				AKSWebhooksSpec(ctx, func() AKSWebhooksSpecInput {
+					return AKSWebhooksSpecInput{
+						Cluster:       result.Cluster,
+						MachinePools:  result.MachinePools,
+						WaitIntervals: e2eConfig.GetIntervals(specName, "wait-machine-pool-nodes"),
+					}
+				})
+			})
+
 			By("Upgrading the Kubernetes version of the cluster", func() {
```
Comment (nojnhuh) [inline test/e2e/azure_test.go:602]: Do we want to keep the upgrade test running first immediately after the cluster is created?

**Review by nojnhuh — COMMENTED** [review]
Left a few small comments inline. This looked like a bit of a slog, so thanks for adding these!

As to whether we should test these e2e at all, I think the fact that you caught at least some unexpected behavior here is a decent indicator that unit tests aren't going to cover everything. If we can get the same benefits in narrower tests with something like envtest, then that might be worth looking into, but I don't think that should block this.

File: test/e2e/aks.go [hunk 9]
```diff
@@ -208,3 +209,74 @@ func WaitForAKSSystemNodePoolMachinesToExist(ctx context.Context, input WaitForC
...
+	key := client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	azManagedControlPlane := &infrav1exp.AzureManagedControlPlane{}
+	Eventually(func() bool {
+		err := managementClusterClient.Get(ctx, key, azManagedControlPlane)
+		return err == nil
+	}, retryableOperationTimeout, retryableOperationSleepBetweenRetries).Should(BeTrue(), "AzureManagedControlPlane/%s in namespace %s unable to be retrieved from mgmt cluster", name, namespace)
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply test/e2e/aks.go:234]: Didn't run into anything, just biased towards retrying anything that has .0001% I/O error outcomes

[the remark this answers is not in the document]
Comment (jackfrancis) [reply test/e2e/aks.go:234]: I was actually going to submit a follow-up PR to add this retry tolerance to all these GET calls :) (including the newly added ones in AKS E2E tests)

File: test/e2e/aks.go [hunk 10]
```diff
@@ -208,3 +209,74 @@ func WaitForAKSSystemNodePoolMachinesToExist(ctx context.Context, input WaitForC
...
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	azCluster := &infrav1.AzureCluster{}
+	err := managementClusterClient.Get(ctx, key, azCluster)
+	return azCluster, err
+}
+
+func getAzureManagedControlPlane(ctx context.Context, managementClusterClient client.Client, namespace, name string) (*infrav1exp.AzureManagedControlPlane, error) {
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply test/e2e/aks.go:224]: Indeed, I was just trying to not add a weird looking func (there are already a bunch w/ a well-established return pattern).

How about this, I'll remove the retry stuff from this PR, and submit a follow-up PR that turns all of these into retryable (and asserts on a terminal error, rather than returns an err).

File: test/e2e/aks_webhooks.go (hunk 5 above)
Comment (jackfrancis) [reply test/e2e/aks_webhooks.go:283]: lol I had PTSD from the `map[string]string` above

File: test/e2e/aks_webhooks.go (hunk 7 above)
Comment (jackfrancis) [reply test/e2e/aks_webhooks.go:379]: I think we probably actually do want to anticipate a test scenario where we have (or update to) > 1 system pool. Because changing back and forth is in fact allowed, we should be cool to do that a bunch of times, and this particular test which does that in an excessive/not allowed way supports that.

(So long as we don't do something weird like leave a million kube-system pods around after we do a multiple system pool scenario and then leave around a bunch of orphaned, non-schedulable pods when we return to only 1 system pool — although even that particular weird edge case should still allow tests tests to pass AFAIK)

File: test/e2e/azure_test.go (hunk 8 above)
Comment (jackfrancis) [reply test/e2e/azure_test.go:602]: I don't think I care, I put these here to get faster test cycles.

**Review by nojnhuh — APPROVED** [review]
/lgtm

File: test/e2e/aks.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (nojnhuh) [reply test/e2e/aks.go]: IIRC client-go has some level of retry logic built in, so it might be worth looking to see if that's sufficient for us.

**Comment by jackfrancis** [comment]
@CecileRobertMichon I haven't had cycles to look into envtest. I think if we don't feel confident that these tests are improving AKS stability let's just close this PR. Alternatively, if we do think these tests offer some improved protection, then let's add a work item to investigate replacing these w/ envtest in the future. The tests themselves will be easy to remove in the future if we land on a better solution.

My 2¢

**Comment by CecileRobertMichon** [comment]
My vote is to close this and work on adding a less costly test via local apiserver. This adds costs in time and $$ to every PR run (last AKS run took 1.5 hours) which isn't worth the additional coverage this brings IMO.

**Comment by jackfrancis** [comment]
@CecileRobertMichon fair enough!
